### PR TITLE
sql: test namespace entry for multi-region enum is drained properly

### PR DESCRIPTION
**Pull request by arulajmani**
Testing only patch. When the primary region is removed from a
multi-region database we queue up an async job to reclaim the namespace
entry for the multi-region type descriptor (and its array alias). This
patch adds a test to make sure that the namespace entry is drained
properly even if the job runs into an error.

This patch adds a new testing knob to the type schema changer,
RunAfterAllFailOrCancel, to synch on the `OnFailOrCancel` job
completing.

Release note: None

### CODE REVIEW

**Comment by arulajmani** [comment]
As promised on #60437

File: pkg/sql/type_change.go [hunk 1]
```diff
@@ -162,6 +162,9 @@ type TypeSchemaChangerTestingKnobs struct {
 	// RunBeforeEnumMemberPromotion runs before enum members are promoted from
 	// readable to all permissions in the typeSchemaChanger.
 	RunBeforeEnumMemberPromotion func()
+	// RunAfterOnFailOrCancel runs after OnFailOrCancel completes, if
+	// OnFailOrCancel is triggered.
+	RunAfterAllFailOrCancel func()
```
Comment (otan) [inline pkg/sql/type_change.go:167]: do you mean `RunAfterOnFailOrCancel`?

File: pkg/sql/multiregion_test.go [hunk 2]
```diff
@@ -127,3 +130,75 @@ func TestSettingPrimaryRegionAmidstDrop(t *testing.T) {
...
+	defer s.Stopper().Stop(ctx)
+
+	// Setup the test.
+	_, err := sqlDB.Exec(`
+CREATE DATABASE db WITH PRIMARY REGION "us-east-1"; 
+CREATE TABLE db.t(k INT) LOCALITY REGIONAL BY TABLE IN PRIMARY REGION;
+`)
+	require.NoError(t, err)
+
+	_, err = sqlDB.Exec(`ALTER DATABASE db DROP REGION "us-east-1" `)
```
Comment (otan) [inline pkg/sql/multiregion_test.go:182]: nit: extra space

File: pkg/sql/multiregion_test.go [hunk 3]
```diff
@@ -127,3 +130,75 @@ func TestSettingPrimaryRegionAmidstDrop(t *testing.T) {
...
+	params.Knobs.SQLTypeSchemaChanger = &sql.TypeSchemaChangerTestingKnobs{
+		RunBeforeExec: func() error {
+			return errors.New("yikes")
+		},
+		RunAfterAllFailOrCancel: func() {
+			mu.Lock()
+			defer mu.Unlock()
+			numCleanups++
+			if numCleanups == 2 {
+				close(cleanupFinished)
```
Comment (otan) [inline pkg/sql/multiregion_test.go:166]: if `OnFailOrCancel` fails, does it retry, or does the entry get stuck?

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60466)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/sql/multiregion_test.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-MTGPfHQ7XmolybJ7PoV:-MTGPfHQ7XmolybJ7PoW:bhuu70l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c903c36d0a43c21e84bf2590506b533bb29710ae/pkg/sql/multiregion_test.go#L164)):*
> ```Go
> 			mu.Lock()
> 			defer mu.Unlock()
> 			numCleanups++
> ```

Nit: why not make this a bool, and set it after the check?

```
haveWePerformedCleanup := false
...
if !haveWePerformedCleanup {
   haveWePerformedCleanup = true
} else {
   close(cleanupFinished)
}
```

**Review by arulajmani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60466)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/sql/multiregion_test.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-MTGPfHQ7XmolybJ7PoV:-MTH0YtlCZk4Le_XI8OW:b-wyhmx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c903c36d0a43c21e84bf2590506b533bb29710ae/pkg/sql/multiregion_test.go#L164)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: why not make this a bool, and set it after the check?

```
haveWePerformedCleanup := false
...
if !haveWePerformedCleanup {
   haveWePerformedCleanup = true
} else {
   close(cleanupFinished)
}
```
</blockquote></details>

I think thats slightly confusing because when this block is hit, cleanup has already been performed, as the knob runs after cleanup has completed. 

I slightly switched this logic to be explicit about the expectation here, lemme know if that improves the readability. 
___
*[pkg/sql/multiregion_test.go, line 166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-MTEGRy3wxJkiTW5EnrQ-r1-166:-MTGwqpS0a0J-a8qEKmG:bf0w6xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c903c36d0a43c21e84bf2590506b533bb29710ae/pkg/sql/multiregion_test.go#L166)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

if `OnFailOrCancel` fails, does it retry, or does the entry get stuck?
</blockquote></details>

This is exactly the bug I was talking about when we spoke yesterday. https://github.com/cockroachdb/cockroach/issues/60489. This isn't to say things can never be in a bad state -- they can be, if the error is non-retryable, but you can't account for everything :)  

I'll send out a patch once this thing goes in, should be easy enough to test once this testing knob is a thing. I can also update this test to account for that. 
___
*[pkg/sql/type_change.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-MTEGRy3wxJkiTW5EnrR-r1-167:-MTGzmuCAEgo0F857DAp:b-43scv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c903c36d0a43c21e84bf2590506b533bb29710ae/pkg/sql/type_change.go#L167)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

do you mean `RunAfterOnFailOrCancel`?
</blockquote></details>

I'm not even sure how I made that typo tbh

**Review by ajstorm — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60466)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/multiregion_test.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-MTGPfHQ7XmolybJ7PoV:-MTH4YnHB88pOYLzJOdP:bjcri8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c903c36d0a43c21e84bf2590506b533bb29710ae/pkg/sql/multiregion_test.go#L164)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think thats slightly confusing because when this block is hit, cleanup has already been performed, as the knob runs after cleanup has completed. 

I slightly switched this logic to be explicit about the expectation here, lemme know if that improves the readability. 
</blockquote></details>

I like the changes.  You could still go with the bool if you change name to `haveWePerformedFirstRoundOfCleanup`, but your call.

**Review by otan — APPROVED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60466)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/type_change.go, line 167 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-MTEGRy3wxJkiTW5EnrR-r1-167:-MTHiS7m400vgX6Ifpm6:bstnkqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c903c36d0a43c21e84bf2590506b533bb29710ae/pkg/sql/type_change.go#L167)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I'm not even sure how I made that typo tbh
</blockquote></details>

you need sleep

**Review by arulajmani — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-:-MTpw7hWBDfG56fihOt3:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/sql/multiregion_test.go, line 164 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60466#-MTGPfHQ7XmolybJ7PoV:-MTpw3kpEwVJ7g33m9EN:b-896fja) ([raw file](https://github.com/cockroachdb/cockroach/blob/c903c36d0a43c21e84bf2590506b533bb29710ae/pkg/sql/multiregion_test.go#L164)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

I like the changes.  You could still go with the bool if you change name to `haveWePerformedFirstRoundOfCleanup`, but your call.
</blockquote></details>

Done!

**Pull request merged**